Test test_https_warnings is not functional with pytest-httpbin 1.0

test_https_warnings is expecting the certificates from pytest-httpbin to produce SubjectAltNameWarning.

pytest-httpbin 1.0 was released, but it has newer certificates that do not produce the warning:

https://github.com/kevin1024/pytest-httpbin/pull/53

And hence the version of pytest-httpbin is pinned here:

https://github.com/psf/requests/blob/1b417634721ec377abb7f17bc1f215e07202c2f7/Pipfile#L16

However with old certificates from pytest-httpbin \< 1.0, we are [no longer unable to test requests on recent Fedora](https://bugzilla.redhat.com/show_bug.cgi?id=1676023), because we get:

    pytest-httpbin server hit an exception serving request: [SSL: CA_MD_TOO_WEAK] ca md too weak (_ssl.c:4020)

Hence we need to update pytest-httpbin in Fedora to 1.0+ in order to test requests, but that way, test_https_warnings fails.

## Expected Result

test_https_warnings should pass with recent pytest-httpbin, possibly by bundling its own certificates.

## Actual Result

test_https_warnings fails:

```
=================================== FAILURES ===================================
_______________________ TestRequests.test_https_warnings _______________________

self = <tests.test_requests.TestRequests object at 0x7fc802ebaa60>
httpbin_secure = <function prepare_url.<locals>.inner at 0x7fc8031980d0>
httpbin_ca_bundle = None

    def test_https_warnings(self, httpbin_secure, httpbin_ca_bundle):
        """warnings are emitted with requests.get"""
        if HAS_MODERN_SSL or HAS_PYOPENSSL:
            warnings_expected = ('SubjectAltNameWarning', )
        else:
            warnings_expected = ('SNIMissingWarning',
                                 'InsecurePlatformWarning',
                                 'SubjectAltNameWarning', )
    
        with pytest.warns(None) as warning_records:
            warnings.simplefilter('always')
            requests.get(httpbin_secure('status', '200'),
                         verify=httpbin_ca_bundle)
    
        warning_records = [item for item in warning_records
                           if item.category.__name__ != 'ResourceWarning']
    
        warnings_category = tuple(
            item.category.__name__ for item in warning_records)
>       assert warnings_category == warnings_expected
E       AssertionError: assert () == ('SubjectAltNameWarning',)
E         Right contains one more item: 'SubjectAltNameWarning'
E         Full diff:
E         - ()
E         + ('SubjectAltNameWarning',)

tests/test_requests.py:855: AssertionError
----------------------------- Captured stderr call -----------------------------
127.0.0.1 - - [11/Jul/2020 00:24:33] "GET /status/200 HTTP/1.1" 200 0
======== 1 failed, 526 passed, 12 skipped, 1 xfailed in 175.72 seconds =========
```

## Reproduction Steps

Unpin the pytest-httpbin version, run tox.

```
$ tox -e py38
$ .tox/py38/bin/python -m pip install -U pytest-httpbin pygments
$ tox -e py38
```

## System Information

    $ python -m requests.help

```
{
  "chardet": {
    "version": "3.0.4"
  },
  "cryptography": {
    "version": ""
  },
  "idna": {
    "version": "2.10"
  },
  "implementation": {
    "name": "CPython",
    "version": "3.8.3"
  },
  "platform": {
    "release": "5.7.6-201.fc32.x86_64",
    "system": "Linux"
  },
  "pyOpenSSL": {
    "openssl_version": "",
    "version": null
  },
  "requests": {
    "version": "2.24.0"
  },
  "system_ssl": {
    "version": "1010107f"
  },
  "urllib3": {
    "version": "1.25.9"
  },
  "using_pyopenssl": false
}
```

--- a/requirements-dev.txt
+++ b/requirements-dev.txt
@@ -1,7 +1,8 @@
 pytest>=2.8.0,<=3.10.1
 pytest-cov
-pytest-httpbin<1.0
+pytest-httpbin==1.0.0
 pytest-mock==2.0.0
 httpbin==0.7.0
 Flask>=1.0,<2.0
+trustme
 wheel

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -1,7 +1,19 @@
 # -*- coding: utf-8 -*-
 
+try:
+    from http.server import HTTPServer
+    from http.server import SimpleHTTPRequestHandler
+except ImportError:
+    from BaseHTTPServer import HTTPServer
+    from SimpleHTTPServer import SimpleHTTPRequestHandler 
+
+import ssl
+import tempfile
+import threading
+
 import pytest
 from requests.compat import urljoin
+import trustme
 
 
 def prepare_url(value):
@@ -22,3 +34,25 @@ def httpbin(httpbin):
 @pytest.fixture
 def httpbin_secure(httpbin_secure):
     return prepare_url(httpbin_secure)
+
+
+@pytest.fixture
+def nosan_server(tmp_path_factory):
+    tmpdir = tmp_path_factory.mktemp("certs")
+    ca = trustme.CA()
+    # only commonName, no subjectAltName
+    server_cert = ca.issue_cert(common_name=u"localhost")
+    ca_bundle = str(tmpdir / "ca.pem")
+    ca.cert_pem.write_to_path(ca_bundle)
+
+    context = ssl.create_default_context(ssl.Purpose.CLIENT_AUTH)
+    server_cert.configure_cert(context)
+    server = HTTPServer(("localhost", 0), SimpleHTTPRequestHandler)
+    server.socket = context.wrap_socket(server.socket, server_side=True)
+    server_thread = threading.Thread(target=server.serve_forever)
+    server_thread.start()
+
+    yield "localhost", server.server_address[1], ca_bundle
+
+    server.shutdown()
+    server_thread.join()

--- a/tests/test_requests.py
+++ b/tests/test_requests.py
@@ -874,8 +874,9 @@ class TestRequests:
         r = requests.get(httpbin(), cert='.')
         assert r.status_code == 200
 
-    def test_https_warnings(self, httpbin_secure, httpbin_ca_bundle):
+    def test_https_warnings(self, nosan_server):
         """warnings are emitted with requests.get"""
+        host, port, ca_bundle = nosan_server
         if HAS_MODERN_SSL or HAS_PYOPENSSL:
             warnings_expected = ('SubjectAltNameWarning', )
         else:
@@ -885,8 +886,7 @@ class TestRequests:
 
         with pytest.warns(None) as warning_records:
             warnings.simplefilter('always')
-            requests.get(httpbin_secure('status', '200'),
-                         verify=httpbin_ca_bundle)
+            requests.get("https://localhost:{}/".format(port), verify=ca_bundle)
 
         warning_records = [item for item in warning_records
                            if item.category.__name__ != 'ResourceWarning']

